File: Dukan_x/lib/features/mobile_shop/config/retry_config.dart
/// Retry and Backoff Configuration (Flutter)
///
/// Defines retry budgets, base/max delays, jitter, and throttling behavior
/// for client-side operations (sync push, provider calls, WebSocket reconnect).
///
/// Requirement: 6.38, 7.2–7.3
library;

import 'package:flutter/foundation.dart';

/// A single retry policy.
@immutable
class RetryPolicy {
  /// Maximum retry attempts before giving up.
  final int maxRetries;

  /// Initial delay before first retry.
  final Duration baseDelay;

  /// Maximum delay cap.
  final Duration maxDelay;

  /// Jitter factor (0–1): randomness added to prevent thundering herd.
  final double jitterFactor;

  /// Backoff multiplier per attempt (exponential backoff).
  final double backoffMultiplier;

  const RetryPolicy({
    required this.maxRetries,
    required this.baseDelay,
    required this.maxDelay,
    required this.jitterFactor,
    required this.backoffMultiplier,
  });
}

/// Retry configuration for all Flutter client operations.
@immutable
class RetryConfig {
  /// Sync push retries (outbox → backend).
  final RetryPolicy syncPush;

  /// External provider call retries.
  final RetryPolicy providerCall;

  /// WebSocket reconnection retries.
  final RetryPolicy websocketReconnect;

  /// API request retries (general).
  final RetryPolicy apiRequest;

  const RetryConfig({
    required this.syncPush,
    required this.providerCall,
    required this.websocketReconnect,
    required this.apiRequest,
  });
}

/// Default retry configuration.
const kRetryConfig = RetryConfig(
  syncPush: RetryPolicy(
    maxRetries: 5,
    baseDelay: Duration(seconds: 1),
    maxDelay: Duration(seconds: 60),
    jitterFactor: 0.25,
    backoffMultiplier: 2.0,
  ),
  providerCall: RetryPolicy(
    maxRetries: 3,
    baseDelay: Duration(milliseconds: 200),
    maxDelay: Duration(seconds: 5),
    jitterFactor: 0.2,
    backoffMultiplier: 2.0,
  ),
  websocketReconnect: RetryPolicy(
    maxRetries: 10,
    baseDelay: Duration(seconds: 1),
    maxDelay: Duration(seconds: 60),
    jitterFactor: 0.5,
    backoffMultiplier: 1.5,
  ),
  apiRequest: RetryPolicy(
    maxRetries: 3,
    baseDelay: Duration(milliseconds: 500),
    maxDelay: Duration(seconds: 10),
    jitterFactor: 0.25,
    backoffMultiplier: 2.0,
  ),
);

File: Dukan_x/my-backend/src/modules/mobile-shop/config/retry.config.ts
/**
 * Retry and Backoff Configuration
 *
 * Defines retry budgets, base/max delays, jitter, and throttling behavior.
 * After the budget is exhausted the system returns a typed rate-limited or
 * pending-reconciliation outcome and preserves idempotency/authoritative state.
 *
 * Requirements: 6.38
 */

export interface RetryPolicy {
  /** Maximum retry attempts before giving up */
  readonly maxRetries: number;
  /** Initial delay before first retry (ms) */
  readonly baseDelayMs: number;
  /** Maximum delay cap (ms) */
  readonly maxDelayMs: number;
  /** Jitter factor (0–1): randomness added to prevent thundering herd */
  readonly jitterFactor: number;
  /** Backoff multiplier per attempt (exponential backoff) */
  readonly backoffMultiplier: number;
}

export interface RetryConfig {
  /** DynamoDB transact/write retries on throttle or transient error */
  readonly dynamoDbWrite: RetryPolicy;
  /** DynamoDB read retries (query/get) */
  readonly dynamoDbRead: RetryPolicy;
  /** Reconciliation worker step retries */
  readonly reconciliationStep: RetryPolicy;
  /** Sync push retries (Flutter outbox) */
  readonly syncPush: RetryPolicy;
  /** External provider call retries */
  readonly providerCall: RetryPolicy;
  /** WebSocket reconnection retries */
  readonly websocketReconnect: RetryPolicy;
  /** Migration/backfill page retries */
  readonly migrationPage: RetryPolicy;
}

export const RETRY_CONFIG: RetryConfig = {
  dynamoDbWrite: {
    maxRetries: 3,
    baseDelayMs: 100,
    maxDelayMs: 2000,
    jitterFactor: 0.25,
    backoffMultiplier: 2,
  },

  dynamoDbRead: {
    maxRetries: 2,
    baseDelayMs: 50,
    maxDelayMs: 1000,
    jitterFactor: 0.1,
    backoffMultiplier: 2,
  },

  reconciliationStep: {
    maxRetries: 5,
    baseDelayMs: 500,
    maxDelayMs: 30_000,
    jitterFactor: 0.3,
    backoffMultiplier: 2,
  },

  syncPush: {
    maxRetries: 5,
    baseDelayMs: 1000,
    maxDelayMs: 60_000,
    jitterFactor: 0.25,
    backoffMultiplier: 2,
  },

  providerCall: {
    maxRetries: 3,
    baseDelayMs: 200,
    maxDelayMs: 5000,
    jitterFactor: 0.2,
    backoffMultiplier: 2,
  },

  websocketReconnect: {
    maxRetries: 10,
    baseDelayMs: 1000,
    maxDelayMs: 60_000,
    jitterFactor: 0.5,
    backoffMultiplier: 1.5,
  },

  migrationPage: {
    maxRetries: 3,
    baseDelayMs: 200,
    maxDelayMs: 5000,
    jitterFactor: 0.2,
    backoffMultiplier: 2,
  },
} as const;
